dolphin/api/v1/alert.py

```diff
@@ -0,0 +1,158 @@
+# Copyright 2020 The SODA Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from oslo_log import log
+from webob import exc
+
+from dolphin import db, cryptor
+from dolphin import exception
+from dolphin import utils
+from dolphin.api import validation
+from dolphin.api.common import wsgi
+from dolphin.api.schemas import alert as schema_alert
+from dolphin.api.views import alert as alert_view
+from dolphin.drivers import api as driver_api
+from dolphin.i18n import _
+
+LOG = log.getLogger(__name__)
+
+
+SNMPv3_keys = ('username', 'password', 'auth_enable', 'auth_protocol', 'encryption_enable',
+               'encryption_protocol', 'encryption_password', 'engine_id')
+
+
+class AlertController(wsgi.Controller):
+    def __init__(self):
+        super().__init__()
+        self.driver_api = driver_api.API()
+
+    @staticmethod
+    def exist(ctx, storage_id):
+        try:
+            db.alert_source_get(ctx, storage_id)
+        except exception.AlertSourceNotFound:
+            return False
+
+        return True
+
+    @staticmethod
+    def decrypt_password(alert_source):
+        password = alert_source.get('password', None)
+        encryption_password = alert_source.get('encryption_password', None)
+        if password:
+            alert_source['password'] = cryptor.decode(password)
+        if encryption_password:
+            alert_source['encryption_password'] = cryptor.decode(encryption_password)
+
+        return alert_source
+
+    @wsgi.response(200)
+    @validation.schema(schema_alert.update)
+    def update(self, req, id, body):
+        """Create a new alert source or update an exist one."""
```

Comment:
IMO, update should only update and not create. If the source is not found, return back proper message and let the caller explicitly ask for create

Reply:
_RFC-2616 clearly mention that PUT method requests for the enclosed entity be stored under the supplied Request-URI. If the Request-URI refers to an already existing resource – an update operation will happen, otherwise create operation should happen if Request-URI is a valid resource URI._  -- refered from https://restfulapi.net/rest-put-vs-post/. So I just renamed the function name from update to put. And the function is used to handle the PUT request.